### DISCUSSION TITLE
How do I automatically add a plugin in a `config.nu` ?

### DISCUSSION THREAD

**Discussion by crabdancing** [nushell/nushell#12997] [Q&A] [answered]
I use a shared `config.nu` between multiple systems & multiple accounts, and I find for every single one, I have to explicitly run `plugin add` for it to take -- it doesn't seem to work inside of `config.nu`, even though the code is definitely being registered. How do I add a plugin automatically?

**Comment by fdncred** [comment]
Plugins have to be added to the `$nu.plugin-path` file with registration via the `plugin add` command. Technically, you may be able to copy that file from system to system as long as the paths and plugins were identical, but that's kind of an unsupported solution. Once a plugin is added to the plugin registry you can run `plugin use` to make that plugin available, or just quit out and restart nushell.

**Reply by crabdancing** [reply]
Hmm.
I'm confused -- why doesn't `plugin add` work from `config.nu`? Is there any context in which I can run the command automatically, without having to resort to calling nushell externally and piping the command in?

**Comment by devyn** [comment]
Installed plugins are kind of meant to be a locally registered thing, because they depend on the absolute path and also the metadata associated with that specific version of the plugin that's installed. You do not have to use `plugin add` or `plugin use` in your config file to make plugins available though, any plugins you have previously registered are already available every time you run `nu` without editing your config file at all.

We don't really recommend doing `plugin add` inside of your config because that's generally going to be slow. It has to launch each plugin to get its commands and then add them to the plugin signature file every time you launch `nu`.

What I recommend instead is that you set up a separate script that installs your plugins and calls `plugin add` to register them in your plugin registry file that you can run when using a machine for the first time. Then you can just run e.g. `nu install-plugins.nu` on each of your machines once.

After calling `plugin add`, you can use `plugin use` to bring them into scope, but this will not work in a script file or in your `config.nu` file.

**Reply by crabdancing** [reply]
> What I recommend instead is that you set up a separate script that installs your plugins and calls plugin add to register them in your plugin registry file that you can run when using a machine for the first time. Then you can just run e.g. nu install-plugins.nu on each of your machines once.

Oh, cool! Could you give an example of how to do this? :)

**Reply by devyn** [reply]
Sure, I would just go for something simple like

```nushell
cargo install nu_plugin_dbus
cargo install nu_plugin_emoji
glob ~/.cargo/bin/nu_plugin_* |
  each { |name|
    print -e $">> (ansi cyan)($name)(ansi reset)"
    try {
      plugin add $name
    } catch { |err|
        print -e $"(ansi red)($err.msg)(ansi reset)"
    }
  }
```

Adjust it to your purposes - probably the glob to find and add plugins is the most helpful part

**Reply by crabdancing** [reply, marked as answer]
Thanks for the input! I figured out a Nix expression I like, which generates the msgpackz file declaratively in the Nix store:

```nix
  activateNushellPluginsNuScript = pkgs.writeTextFile {
    name = "activateNushellPlugins";
    destination = "/bin/activateNushellPlugins.nu";
    text = ''
      #!/usr/bin/env nu
      ${
        concatStringsSep "\n" (map (x: "plugin add ${nushellPlugins.${x}}/bin/nu_plugin_${x}") pluginNamesNixpkgs)
      }
    '';
  };

   msgPackz =
    pkgs.runCommand "nushellMsgPackz" {}
    ''
      mkdir -p "$out"
      # After some experimentation, I determined that this only works if --plugin-config is FIRST
      ${pkgs.nushell}/bin/nu --plugin-config "$out/plugin.msgpackz" ${activateNushellPluginsNuScript}/bin/activateNushellPlugins.nu
    '';

```

It can then be symlinked with home manager against `~/.config/nushell/plugin.msgpackz`:

```nix
    xdg.configFile."nushell/plugin.msgpackz".source = "${msgPackz}/plugin.msgpackz";
```

The nice thing about this approach is that it's fully reproducible & ideopotent, while still caching the results. :)

If I want to, I can also wrap nushell to point at the file generated in the derivation, thus ensuring the plugin versions are reliably mapped to the nushell version, so that environment config doesn't break older/newer versions of nushell that are being loaded up ephemerally.

It's not perfect, but I'm happy with my setup now.
